docs/plugins/plugin-writer/first-plugin.rst

```diff
@@ -1,2 +1,106 @@
+ * :class:`pulpcore.plugin.models.Importer` should be subclassed and extended with additional attributes to the plugin needs,
+ * define ``TYPE`` class attribute which is used for filtering purposes,
+ * ``sync`` method should be defined on a plugin importer model ``ExampleImporter``,
+ * create a serializer for your new importer as a subclass of :class:`pulpcore.plugin.serializers.ImporterSerializer`,
+ * create a viewset for your new importer as a subclass of :class:`pulpcore.plugin.viewsets.ImporterViewSet`.
+
+:class:`~pulpcore.plugin.models.Importer` model should not be used directly anywhere in plugin code. Only plugin-defined Importer classes are expected to be used.
+
+One of the ways to perform synchronization:
+ * Download and analyze repository metadata from a remote source.
+ * Decide what needs to be added to repository or removed from it.
+ * Associate already existing content to a repository by creating an instance of :class:`~pulpcore.plugin.models.RepositoryContent` and saving it.
+ * Remove :class:`~pulpcore.plugin.models.RepositoryContent` objects which were identified for removal.
+ * For every content which should be added to Pulp create but do not save yet:
+
+  * instance of ``ExampleContent`` which will be later associated to a repository.
+  * instance of :class:`~pulpcore.plugin.models.ContentArtifact` to be able to create relations with the artifact models.
+  * instance of :class:`~pulpcore.plugin.models.RemoteArtifact` to store information about artifact from remote source and to make a relation with  :class:`~pulpcore.plugin.models.ContentArtifact` created before.
+
+ * If a remote content should be downloaded right away (aka ``immediate`` download policy), use one of the suggested download solutions (:ref:`concurrent.futures <futures-docs>` or :ref:`asyncio <asyncio-docs>` approach) to start downloading. If content should be downloaded later (aka ``on_demand`` or ``background`` download policy), feel free to skip this step.
+ * Save all artifact and content data in one transaction:
+
+  * in case of downloaded content, create instance of :class:`~pulpcore.plugin.models.Artifact` which refers to a downloaded file on a filesystem and contains calculated checksums for it.
+  * in case of downloaded content, update :class:`~pulpcore.plugin.models.ContentArtifact` with reference to the created :class:`~pulpcore.plugin.models.Artifact`.
+  * create and save instance of :class:`~pulpcore.plugin.models.RepositoryContent` to associate content to a repository.
+  * save all created artifacts and content: ``ExampleContent``, :class:`~pulpcore.plugin.models.ContentArtifact`, :class:`~pulpcore.plugin.models.RemoteArtifact`.
+
+ * Use :class:`~pulpcore.plugin.models.ProgressBar` to report progress of some steps if needed.
+
+Check our Example plugin implementation of importers: `one uses asyncio <link to pulp_example>`_ as a solution for downloading, `the other one follows concurrent.futures approach <link to pulp_example>`_.
+
```

Comment:
I'm concerned that this does not include the File plugin which uses the `ChangeSet`.  The ExamplePlugin does a good job at illustrating all the details of what needs to be done but I really think we should be encouraging plugin writers to use the `ChangeSet`.

Reply:
Added